resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/StagedResourcesExpirationManager.scala

```diff
@@ -0,0 +1,121 @@
+      clock: Clock,
+      expiredResourceTtlMs: Long,
+      resourceCleanupIntervalMs: Long)
+    extends StagedResourcesExpirationManager {
+
+  private val activeResources = TrieMap.empty[String, MonitoredResource]
+
+  override def startMonitoringForExpiredResources(): Unit = {
+    expirationExecutorService.scheduleAtFixedRate(
+        new CleanupRunnable(),
+        resourceCleanupIntervalMs,
+        resourceCleanupIntervalMs,
+        TimeUnit.MILLISECONDS)
+  }
+
+  override def monitorResourceForExpiration(
+      resourceId: String,
+      podNamespace: String,
+      podLabels: Map[String, String],
+      kubernetesCredentials: PodMonitoringCredentials): Unit = {
+    activeResources(resourceId) = MonitoredResource(
+        resourceId, podNamespace, podLabels, kubernetesCredentials)
+  }
+
+  private class CleanupRunnable extends Runnable with Logging {
+
+    private val expiringResources = mutable.Map.empty[String, ExpiringResource]
+
+    override def run(): Unit = {
+      // Make a copy so we can iterate through this while modifying activeResources
+      val activeResourcesCopy = Map.apply(activeResources.toSeq: _*)
+      for ((resourceId, resource) <- activeResourcesCopy) {
+        Utils.tryWithResource(
+            kubernetesClientProvider.getKubernetesClient(
+                resource.podNamespace, resource.kubernetesCredentials)) { kubernetesClient =>
+          val podsWithLabels = kubernetesClient
+            .pods()
+            .withLabels(resource.podLabels.asJava)
+            .list()
+          if (podsWithLabels.getItems.isEmpty) {
+            activeResources.remove(resourceId)
+            val expiresAt = clock.getTimeMillis() + expiredResourceTtlMs
+            expiringResources(resourceId) = ExpiringResource(expiresAt, resource)
+          }
+        }
+      }
+
+      val expiringResourcesCopy = Map.apply(expiringResources.toSeq: _*)
+      for ((resourceId, expiringResource) <- expiringResourcesCopy) {
+        Utils.tryWithResource(
+            kubernetesClientProvider.getKubernetesClient(
+                expiringResource.resource.podNamespace,
+                expiringResource.resource.kubernetesCredentials)) { kubernetesClient =>
+          val podsWithLabels = kubernetesClient
+            .pods()
+            .withLabels(expiringResource.resource.podLabels.asJava)
+            .list()
+          if (!podsWithLabels.getItems.isEmpty) {
+            activeResources(resourceId) = expiringResource.resource
+            expiringResources.remove(resourceId)
+          } else if (clock.getTimeMillis() > expiringResource.expiresAt) {
+            stagedResourcesStore.removeResources(resourceId)
+            expiringResources.remove(resourceId)
+          }
+        }
+      }
+    }
+  }
+
+  private case class MonitoredResource(
+      resourceId: String,
+      podNamespace: String,
+      podLabels: Map[String, String],
+      kubernetesCredentials: PodMonitoringCredentials)
```

Comment:
The goal here was to make it such that this doesn't have to use a service account and just uses the credentials that are provided by the users. But I'm not sure if this is the correct approach as it complicates the API.